cdap-app-fabric/src/main/java/io/cdap/cdap/internal/metadata/MetadataConsumerSubscriberService.java

```diff
@@ -0,0 +1,305 @@
+        processor.processMessage(message, structuredTableContext);
+        conflictCount = 0;
+      } catch (ConflictException e) {
+        if (messageId.equals(conflictMessageId)) {
+          conflictCount++;
+          if (conflictCount >= maxRetriesOnConflict) {
+            LOG.warn("Skipping metadata message {} after processing it has caused {} consecutive conflicts: {}",
+                     message, conflictCount, e.getMessage());
+            continue;
+          }
+        } else {
+          conflictMessageId = messageId;
+          conflictCount = 1;
+        }
+        throw e;
+      }
+    }
+  }
+
+  /**
+   * The {@link MetadataMessageProcessor} for processing field lineage.
+   */
+  private static final class FieldLineageProcessor implements MetadataMessageProcessor {
+
+    private final CConfiguration cConf;
+    private final Map<String, MetadataConsumer> consumers;
+
+    FieldLineageProcessor(CConfiguration cConf, Map<String, MetadataConsumer> consumers) {
+      this.cConf = cConf;
+      this.consumers = consumers;
+    }
+
+    @Override
+    public void processMessage(MetadataMessage message, StructuredTableContext context) {
+      if (!(message.getEntityId() instanceof ProgramRunId)) {
+        LOG.warn("Missing program run id from the field lineage information. Ignoring the message {}", message);
+        return;
+      }
+
+      ProgramRunId programRunId = (ProgramRunId) message.getEntityId();
+      FieldLineageInfo fieldLineageInfo;
+      try {
+        Gson gson = new GsonBuilder()
+          .registerTypeAdapter(EntityId.class, new EntityIdTypeAdapter())
+          .registerTypeAdapter(MetadataOperation.class, new MetadataOperationTypeAdapter())
+          .registerTypeAdapter(Operation.class, new OperationTypeAdapter())
+          .registerTypeAdapter(EndPointField.class, new EndpointFieldDeserializer())
+          .create();
+
+        fieldLineageInfo = message.getPayload(gson, FieldLineageInfo.class);
+      } catch (Throwable t) {
+        LOG.warn("Error while deserializing the field lineage information message received from TMS. Ignoring : {}",
+                 message, t);
+        return;
+      }
+      // create ProgramRun and LineageInfo for MetadataConsumer
+      long startTimeMs = RunIds.getTime(programRunId.getRun(), TimeUnit.MILLISECONDS);
+      long endTimeMs = System.currentTimeMillis();
+      ProgramRun run = getProgramRunForConsumer(programRunId, startTimeMs, endTimeMs);
+      LineageInfo info = getLineageInfoForConsumer(fieldLineageInfo, startTimeMs, endTimeMs);
+      this.consumers.forEach((key, consumer) -> {
+        try {
+          DefaultMetadataConsumerContext metadataConsumerContext =
+            new DefaultMetadataConsumerContext(cConf, consumer.getName());
+          consumer.consumeLineage(metadataConsumerContext, run, info);
+        } catch (Throwable e) {
+          LOG.error("Error initializing metadata consumer {}: {}", consumer.getName(), e.getMessage());
```

Comment:
Is logging and continuing on failure the expected behavior? If so, add a comment. nit: fix the message since it is not only initializing.

Reply:
That's right. Added a comment